lib/js/jsffi.nim

```diff
@@ -0,0 +1,382 @@
+    ## Dynamically typed wrapper around a JavaScript object.
+  JsAssoc*[K, V] = ref object of JsRoot
+    ## Statically typed wrapper around a JavaScript object.
+
+# Import variables:
+template magicVar*(name, typ: untyped): untyped =
+  ## Imports a variable declared in JavaScript code. This is useful for
+  ## importing globals or things like `self` or `this` from JavaScript.
+  ## It takes the name of the variable and its type.
+  ##
+  ## Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  # We want to import the console object as an JsObject
+  ##  magicVar(console, JsObject)
+  ##  console.log("Hello JavaScript!")
+  ##
+  var `name` {. importc, nodecl, inject .}: `typ`
+
+# New
+proc newJsObject*: JsObject {. importcpp: "{@}" .}
+  ## Creates a new empty JsObject
+proc newJsAssoc*[K, V]: JsAssoc[K, V] = {. importcpp: "{@}" .}
+  ## Creates a new empty JsAssoc with key type `K` and value type `V`.
+
+# Checks
+proc hasOwnProperty*(x: JsObject, prop: string): bool
+  {. importcpp: "#.hasOwnProperty(#)" .}
+  ## Checks, whether `x` has a property of name `prop`.
+
+proc jsTypeOf*(x: JsObject): cstring {. importcpp: "typeof(#)" .}
+  ## Returns the name of the JsObject's JavaScript type as a cstring.
+
+# Conversion to and from JsObject
+proc to*(x: JsObject, T: typedesc): T {. importcpp: "(#)" .}
+  ## Converts a JsObject `x` to type `T`.
+proc toJs*[T](val: T): JsObject {. importcpp: "(#)" .}
+  ## Converts a value of any type to type JsObject
+
+proc callFieldImpl(obj, field, obj1: JsObject,
+  args: varargs[JsObject]): JsObject {. importcpp: callImpl, discardable .}
+
+proc `[]`*(obj: JsObject, field: cstring): JsObject {. importcpp: getImpl .}
+  ## Return the value of a property of name `field` from a JsObject `obj`.
+
+proc `[]=`*[T](obj: JsObject, field: cstring, val: T) {. importcpp: setImpl .}
+  ## Set the value of a property of name `field` in a JsObject `obj` to `v`.
+
+proc `[]`*[K, V](obj: JsAssoc[K, V], field: K): V {. importcpp: getImpl .}
+  ## Return the value of a property of name `field` from a JsAssoc `obj`.
+
+proc `[]`*[V](obj: JsAssoc[string, V], field: string): V
+  {. importcpp: getImplJsString .}
+  ## Return the value of a property of name `field` from a JsAssoc `obj`.
+
+proc `[]=`*[K, V](obj: JsAssoc[K, V], field: K, val: V) {. importcpp: setImpl .}
+  ## Set the value of a property of name `field` in a JsAssoc `obj` to `v`.
+
+proc `[]=`*[V](obj: JsAssoc[string, V], field: string, val: V)
+  {. importcpp: setImplJsString .}
+  ## Set the value of a property of name `field` in a JsAssoc `obj` to `v`.
+
+proc `==`*(x, y: JsRoot): bool {. importcpp: "(# === #)" .}
+  ## Compare two JsObjects or JsAssocs. Be careful though, as this is comparison
+  ## like in JavaScript, so if your JsObjects are in fact JavaScript Objects,
+  ## and not strings or numbers, this is a *comparison of references*.
+
+{. experimental .}
+proc `.`*(obj: JsObject, field: cstring): JsObject {. importcpp: getImpl .}
```

Comment:
I'm a bit reluctant to use dot-notation here, as in nim there are a variety of valid field names (`if`, `for`), which cause dot-notation to break for old versions of JavaScript (like version 1.5, [which nim is supposed to emit](http://nim-lang.org/docs/backends.html#backends-the-javascript-target)), or even all versions (`==`, and other kinds of operator-like identifiers).
So even if those field names are somewhat unorthodox, I think they should be supported.